src/ccl_halomod.c

```diff
@@ -0,0 +1,300 @@
+// Alternatively - just use this in ccl_massfunc as necessary
+double nu(ccl_cosmology *cosmo, double halomass, double a, int * status) {
+  return delta_c()/ccl_sigmaM(cosmo, halomass, a, status);
+}
+
+// TODO: Actually write this
+double z_formation_Bullock(ccl_cosmology *cosmo, double halomass, double a, int * status){
+  // Halo "formation redshift" calculated according to the Bullock et al. (2001) prescription
+  return 0.;
+}
+
+// TODO: make consistency check so that ccl_halo_concentration only runs if called with appropriate definition
+// TODO: should this be moved to the ccl_massfunc.c ?
+// TODO: e.g. if Delta != 200 rho_{mean}, should not function (or should it?)
+double ccl_halo_concentration(ccl_cosmology *cosmo, double halomass, double a, int * status)
+{
+  // Set concentration-mass relation
+  // 1 - Bhattaharya et al. (2011)
+  // 2 - Bullock et al. (2001)
+  // 3 - Duffy et al. (2008)
+  // 4 - Constant concentration (for testing)
+  int iconc=1;
+
+  // Bhattacharya et al. 2011, Delta = 200 rho_{mean} (Table 2)
+  if(iconc==1){    
+    double gz = ccl_growth_factor(cosmo,a,status);
+    double g0 = ccl_growth_factor(cosmo,1.0,status);
+    //return 9.*pow(nu(cosmo,halomass,a,status),-.29)*pow(ccl_growth_factor(cosmo, a, status)/ccl_growth_factor(cosmo, 1.0, status),1.15)
+    return 9.*pow(nu(cosmo,halomass,a,status),-0.29)*pow(gz/g0,1.15);
+  }
+
+  // Bullock et al. (2001)
+  else if(iconc==2){    
+    double A = 4.;
+    double z = -1.+1./a;
+    double zf = z_formation_Bullock(cosmo,halomass,a,status);
+    return A*(1.+zf)/(1.+z);
+  }
+
+  // Duffy et al. 2008 (Delta = 200 mean)
+  else if(iconc==3){    
+    return 10.14*pow(halomass/2.e+12,-0.081)*pow(a,1.01); 
+  }
+
+  // Constant concentration (good for tests)
+  else if(iconc==4){
+    return 100.;
+  }
+
+  // Something went wrong
+  else{    
+    exit(0);
+  }
+	  
+}
+
+typedef struct{
+  // Parameters for the I02 integrand
+  ccl_cosmology *cosmo;
+  double k, a;
+  int * status;
+} IntI02Par;
+
+double one_halo_integrand(double log10mass, void *params){  
+  // Integrand for the one-halo integral
+  IntI02Par *p=(IntI02Par *)params;
+  double halomass = pow(10,log10mass);
+
+  // The halo concentration for this mass and scale factor  
+  double c = ccl_halo_concentration(p->cosmo,halomass,p->a,p->status); 
+
+  // The squared normalised Fourier Transform of a halo profile (W(k->0 = 1)
+  double wk_squared = pow(u_nfw_c(p->cosmo,c,halomass,p->k,p->a,p->status),2); 
+
+  // The mean background matter density in Msun/Mpc^3
+  //double rho_matter = RHO_CRITICAL*p->cosmo->params.Omega_m*pow(p->cosmo->params.h,2);
+  double rho_matter = comoving_matter_density(p->cosmo);
+
+  // Fairly sure that there should be no ln(10) factor should be here since the integration is being specified in log10 range
+  double dn_dlogM = ccl_massfunc(p->cosmo,halomass,p->a,Delta_v(),p->status);
+    
+  return pow(halomass,2)*dn_dlogM*wk_squared/pow(rho_matter,2); 
```

Comment:
isn't there a factor of halo mass, and possibly some numeric prefactor missing to convert d M -> d lg_10(M)?

Reply:
I think not in this case, because the integral is done over a range that is set in log10 I think this cancels out in the end. Probably there is a ln(10) that is missing in both a numerator and denominator. I'll think about whether it is better to include this explicitly to make what is going on in the code more transparent...

Reply:
Certainly there is no factor of mass missing. I think the difference is in our definition of the halo profile, where mine carries a mass scaling. So I have W(M,k->0)=M/\bar\rho whereas you have U(M,k->0)=1.